cli/src/cmd/forge/geiger/find.rs

```diff
@@ -0,0 +1,86 @@
+//! scan a file for unsafe usage
+
+use crate::cmd::forge::geiger::error::ScanFileError;
+use forge_fmt::{parse, Visitable, Visitor};
+use foundry_common::fs;
+use serde::{Deserialize, Serialize};
+use solang_parser::{
+    diagnostics::Diagnostic,
+    pt::{Expression, Loc},
+};
+use std::{convert::Infallible, ops::Add, path::Path};
+
+/// Scan a single file for `unsafe` usage.
+pub fn find_cheatcodes_in_file(path: &Path) -> Result<SolFileMetrics, ScanFileError> {
+    let content = fs::read_to_string(path)?;
+    let cheatcodes = find_cheatcodes_in_string(&content)
+        .map_err(|diagnostic| ScanFileError::ParseSol(diagnostic, path.to_path_buf()))?;
+    Ok(SolFileMetrics { cheatcodes })
+}
+
+pub fn find_cheatcodes_in_string(src: &str) -> Result<CheatcodeCounter, Vec<Diagnostic>> {
+    let mut parsed = parse(&src)?;
+    let mut visitor = CheatcodeVisitor::default();
+    let _ = parsed.pt.visit(&mut visitor);
+    Ok(visitor.cheatcodes)
+}
+
+#[derive(Default)]
+struct CheatcodeVisitor {
+    cheatcodes: CheatcodeCounter,
+}
+
+impl Visitor for CheatcodeVisitor {
+    type Error = Infallible;
+
+    fn visit_expr(&mut self, loc: Loc, expr: &mut Expression) -> Result<(), Self::Error> {
+        if let Expression::FunctionCall(loc, lhs, rhs) = expr {
+            dbg!(&lhs);
+        }
+        Ok(())
+    }
+}
+
+/// Scan result for a single `.sol` file.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct SolFileMetrics {
+    /// Cheatcode metrics.
+    pub cheatcodes: CheatcodeCounter,
+}
+
+/// Unsafe usage metrics collection.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct CheatcodeCounter {
+    pub ffi: usize,
+    pub read_file: usize,
+    pub read_line: usize,
+    pub write_file: usize,
+    pub write_line: usize,
+    pub remove_file: usize,
```

Comment:
Thoughts on also adding to this list:
- `setEnv`: I know we're disabling setting FFI with that in #2992, but could be other unsafe env vars (not necessarily foundry related) set with it
- `deriveKey`: It can read a mnemonic from a filepath, not sure if we allow all paths / have that properly scoped

Reply:
good point! will include as well